compiler/rustc_mir_transform/src/jump_threading.rs

```diff
@@ -0,0 +1,449 @@
+
+            if let Some((value, then, else_)) = targets.as_static_if() {
+                let Some(value) = ScalarInt::try_from_uint(value, discr_layout.size) else { continue };
+                state.insert(
+                    discr,
+                    [
+                        Condition { value, polarity: true, target: then },
+                        Condition { value, polarity: false, target: else_ },
+                    ]
+                    .into_iter(),
+                );
+            } else {
+                state.insert(
+                    discr,
+                    targets.iter().filter_map(|(value, target)| {
+                        let value = ScalarInt::try_from_uint(value, discr_layout.size)?;
+                        Some(Condition { value, polarity: true, target })
+                    }),
+                );
+            }
+
+            finder.find_opportunity(bb, state, cost, 0);
+        }
+
+        let opportunities = finder.opportunities;
+        debug!(?opportunities);
+        if opportunities.is_empty() {
+            return;
+        }
+
+        let mut predecessors = predecessor_count(body);
+        for op in opportunities {
+            apply_opportunity(body, op, &mut predecessors);
+        }
+    }
+}
+
+#[derive(Debug)]
+struct ThreadingOpportunity {
+    /// The list of `BasicBlock`s from the one that found the opportunity to the `SwitchInt`.
+    chain: Vec<BasicBlock>,
+    /// The `SwitchInt` will be replaced by `Goto { target }`.
+    target: BasicBlock,
+}
+
+struct TOFinder<'tcx, 'a> {
+    tcx: TyCtxt<'tcx>,
+    param_env: ty::ParamEnv<'tcx>,
+    body: &'a Body<'tcx>,
+    map: Map,
+    opportunities: Vec<ThreadingOpportunity>,
+}
+
+#[derive(Clone)]
+struct ConditionSet<'a> {
+    /// We use an arena to avoid cloning the slices when cloning `state`.
+    arena: &'a DroplessArena,
+    /// Set of conditions that each place can fulfill.
+    state: FxHashMap<PlaceIndex, &'a [Condition]>,
+}
+
+impl std::fmt::Debug for ConditionSet<'_> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        std::fmt::Debug::fmt(&self.state, f)
+    }
+}
+
+impl<'a> ConditionSet<'a> {
+    fn new(arena: &DroplessArena) -> ConditionSet<'_> {
+        ConditionSet { arena, state: FxHashMap::default() }
+    }
+
+    fn is_empty(&self) -> bool {
+        self.state.is_empty()
+    }
+
+    fn insert(&mut self, place: PlaceIndex, conds: impl Iterator<Item = Condition>) {
+        let conds = self.arena.alloc_from_iter(conds);
+        self.state.insert(place, conds);
+    }
+
+    fn insert_ref(&mut self, place: PlaceIndex, conds: &'a [Condition]) {
+        self.state.insert(place, conds);
+    }
+
+    fn take(&mut self, place: PlaceIndex, map: &Map) -> Option<&'a [Condition]> {
+        let current_condition = self.state.remove(&place);
+        map.preorder_invoke(place, &mut |place| {
+            self.state.remove(&place);
+        });
```

Comment:
This does not invalidate enough places for discriminants to be sound. Ideally, a write to an enum field should invalidate all the other variants and the discriminant of the enum recursively, and reciprocally.
This means that `discriminant((_1 as variant#1).0) = ?` invalidates `_1 as other variants` and `((_1 as variant#1).0 as other variants)`.